netplan/cli/commands/set.py

```diff
@@ -0,0 +1,119 @@
+        self.parser.add_argument('--root-dir', default='/',
+                                 help='Overwrite configuration files in this root directory instead of /')
+
+        self.func = self.command_set
+
+        self.parse_args()
+        self.run_command()
+
+    def command_set(self):
+        if len(self.origin_hint) == 0:
+            raise Exception('Invalid/empty origin-hint')
+        split = self.key_value.split('=', 1)
+        key, value = (split[0], None)
+        if len(split) > 1:
+            value = split[1]
+        if key.startswith('network.'):
+            key = key[8:]
+        self.write_file(key, value, self.origin_hint + '.yaml', self.root_dir)
+
+    def parse_key(self, key, value):
+        # Split at '.' but not at '\.' via negative lookbehind expression
+        split = re.split(r'(?<!\\)\.', 'network.' + key)
+        tree = {}
+        i = 1
+        t = tree
+        for part in split:
+            part = part.replace('\\.', '.')  # Unescape interface-ids, containing dots
+            val = {}
+            if i == len(split):
+                val = value
+            t = t.setdefault(part, val)
+            i += 1
+        return tree
+
+    def merge(self, a, b, path=None):
+        "merges b into a"
+        if path is None:
+            path = []
+        for key in b:
+            if key in a:
+                if isinstance(a[key], dict) and isinstance(b[key], dict):
+                    self.merge(a[key], b[key], path + [str(key)])
+                elif b[key] is None:
+                    del a[key]
+                else:
+                    # Overwrite existing key with new key/value from 'set' command
+                    a[key] = b[key]
+            else:
+                a[key] = b[key]
+        return a
+
+    def write_file(self, key, value, name, rootdir='/', path='etc/netplan/'):
+        tmproot = tempfile.TemporaryDirectory(prefix='netplan-set_')
+        os.makedirs(os.path.join(tmproot.name, 'etc', 'netplan'))
+
+        config = {'network': {}}
+        absp = os.path.join(rootdir, path, name)
+        if os.path.isfile(absp):
+            with open(absp, 'r') as f:
+                config = yaml.safe_load(f)
+
+        new_tree = self.merge(config, self.parse_key(key, yaml.safe_load(value)))
+        stripped = ConfigManager.strip_tree(new_tree)
+        if 'network' in stripped:
+            tmpp = os.path.join(tmproot.name, path, name)
+            with open(tmpp, 'w+') as f:
+                new_yaml = yaml.dump(stripped, indent=2, default_flow_style=False)
+                f.write(new_yaml)
+            try:
+                # Validate the newly created file, by parsing it via libnetplan
+                utils.netplan_parse(tmpp)
+                # Valid, move it to final destination
+                os.replace(tmpp, absp)
+            finally:
+                shutil.rmtree(os.path.join(tmproot.name, 'etc', 'netplan'))
```

Comment:
I don't think this is needed. I guess `TemporaryDirectory` objects will get auto-removed when the object is destroyed, as in after an exception or moving out of scope.

Reply:
Correct: "On completion of the context or destruction of the temporary directory object the newly created temporary directory and all its contents are removed from the filesystem."